internal/testrunner/runners/system/runner.go

```diff
@@ -722,36 +1044,83 @@ func (r *runner) runTest(config *testConfig, ctxt servicedeployer.ServiceContext
+			return nil, testrunner.ErrTestCaseFailed{Reason: fmt.Sprintf("the test service %s unexpectedly exited with code %d", config.Service, code)}
 		}
 	}
 
 	if waitErr != nil {
-		return result.WithError(waitErr)
+		return nil, waitErr
 	}
 
 	if !passed {
-		return result.WithError(testrunner.ErrTestCaseFailed{Reason: fmt.Sprintf("could not find hits in %s data stream", dataStream)})
+		// return nil, fmt.Errorf("could not find hits in %s data stream", scenario.dataStream)
+		// result.FailureMsg = fmt.Sprintf("could not find hits in %s data stream", dataStream)
+		// return result.WithError(fmt.Errorf("%s", result.FailureMsg))
+		return nil, testrunner.ErrTestCaseFailed{Reason: fmt.Sprintf("could not find hits in %s data stream", scenario.dataStream)}
 	}
 
 	logger.Debugf("check whether or not synthetics is enabled (component template %s)...", componentTemplatePackage)
-	syntheticEnabled, err := r.isSyntheticsEnabled(dataStream, componentTemplatePackage)
+	scenario.syntheticEnabled, err = r.isSyntheticsEnabled(scenario.dataStream, componentTemplatePackage)
 	if err != nil {
-		return result.WithError(fmt.Errorf("failed to check if synthetic source is enabled: %w", err))
+		return nil, fmt.Errorf("failed to check if synthetic source is enabled: %w", err)
+	}
+	logger.Debugf("data stream %s has synthetics enabled: %t", scenario.dataStream, scenario.syntheticEnabled)
+
+	scenario.docs = hits.getDocs(scenario.syntheticEnabled)
+
+	if r.options.RunSetup {
+		policyBytes, err := json.Marshal(policy)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshall policy: %w", err)
+		}
+
+		origPolicyBytes, err := json.Marshal(origPolicy)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshall policy: %w", err)
+		}
+
+		agentBytes, err := json.Marshal(origAgent)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshall agent: %w", err)
+		}
+
+		testConfigBytes, err := json.Marshal(config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshall test config options: %w", err)
+		}
+
+		err = os.WriteFile(filepath.Join(r.locationManager.ServiceSetupDir(), setupNewPolicyFileName), policyBytes, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write policy JSON: %w", err)
+		}
+		err = os.WriteFile(filepath.Join(r.locationManager.ServiceSetupDir(), setupOrigPolicyFileName), origPolicyBytes, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write origin policy JSON: %w", err)
+		}
+		err = os.WriteFile(filepath.Join(r.locationManager.ServiceSetupDir(), setupAgentFileName), agentBytes, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write agent JSON: %w", err)
+		}
+		err = os.WriteFile(filepath.Join(r.locationManager.ServiceSetupDir(), setupTestConfigFileName), testConfigBytes, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write test config JSON: %w", err)
+		}
```

Comment:
Move all this to another function? And is there a reason why we cannot write all this to a single file?

Reply:
Moved everything to just one file